### PR TITLE
ParticleHandler: account for modified stepping actions, where tracking…

### PR DESCRIPTION
**Pull request by andresailer**
… of a particle could be paused and later restarted.

Pick up the particle information that we have stored and go on from there

BEGINRELEASENOTES
- ParticleHandler: account for modified stepping actions, where tracking of a particle could be paused and later restarted.

ENDRELEASENOTES

For #1036 I get this for 2 events now

```
root [2] EVENT->Scan("sqrt(MCParticles.psx^2 + MCParticles.psy^2 + MCParticles.psz^2 + MCParticles.mass^2):MCParticles.reason:MCParticles.pdgID:MCParticles.mask", "MCParticles.mask==1")
***********************************************************************
*    Row   * Instance * sqrt(MCPa * MCParticl * MCParticl * MCParticl *
***********************************************************************
*        0 *        0 *     10000 *       158 *        11 *         1 *
*        1 *        0 *     10000 *       158 *        11 *         1 *
***********************************************************************
==> 2 selected entries
(long long) 2
```
So the energy is the same as the initial one (10000) as opposed to some smaller number.

### CODE REVIEW

**Comment by andresailer** [comment]
@MarkusFrankATcernch Does this look sensible to you?

**Comment by MarkusFrankATcernch** [comment]
If the tracking action is the first call to the particle handler after the restacking, this is should be correct.
However, I have the suspicion that all properties of the existing "currentTrack" may all somehow lost, because this
track and its properties is not yet saved away to the collection possibly be kept in the output record.

Hence the question is: Whan does the restacking happen? before or after a couple of steps were already processed?

**Comment by MarkusFrankATcernch** [comment]
if before the restacking the member
```
void Geant4ParticleHandler::end(const G4Track* track)
```
is called this approach should be correct. Be aware that this track may be in both collections:
- Geant4ParticleHandler::m_equivalentTracks and
- Geant4ParticleHandler::m_particleMap
in `Geant4ParticleHandler::endEvent(const G4Event* event)` preference must be given to `Geant4ParticleHandler::m_particleMap` when rebasing the tracks.

**Comment by andresailer** [comment]
I implemented the treatment also for particles that might not be in the particleMap before they are suspended. It looks fine to me, but that code is never actually run so far.

**Comment by andresailer** [comment]
I am adding a test

**Comment by andresailer** [comment]
Added a test that exercises the suspendedPM related code

File: DDG4/include/DDG4/Geant4ParticleHandler.h [hunk 1]
```diff
@@ -119,6 +119,9 @@ namespace dd4hep {
       Particle          m_currTrack;
       /// Map with stored MC Particles
       ParticleMap       m_particleMap;
+      /// Map with stored MC Particles that were suspended by the stacking action
```
Comment (andresailer) [inline DDG4/include/DDG4/Geant4ParticleHandler.h:122]:
```suggestion
      /// Map with stored MC Particles that were suspended by the stepping action
```

File: DDG4/src/Geant4ParticleHandler.cpp [hunk 2]
```diff
@@ -115,6 +115,8 @@ bool Geant4ParticleHandler::adopt(Geant4Action* action)    {
 void Geant4ParticleHandler::clear()  {
   detail::releaseObjects(m_particleMap);
   m_particleMap.clear();
+  // m_suspendedPM should already be empty and cleared...
+  assert(m_suspendedPM.empty());
```
Comment (andresailer) [inline DDG4/src/Geant4ParticleHandler.cpp:119]:
```suggestion
  assert(m_suspendedPM.empty() && "There was something wrong with the particle record treatment, please open a bug report!");
```

File: DDG4/src/Geant4ParticleHandler.cpp [hunk 3]
```diff
@@ -293,7 +314,10 @@ void Geant4ParticleHandler::begin(const G4Track* track)   {
 void Geant4ParticleHandler::end(const G4Track* track)   {
   Geant4TrackHandler h(track);
   Geant4ParticleHandle ph(&m_currTrack);
-  int g4_id = h.id();
+  const int g4_id = h.id();
+
+  std::cout << "ending " << h.statusName() << " " << g4_id  << std::endl;
```
Comment (andresailer) [inline DDG4/src/Geant4ParticleHandler.cpp:319]:
```suggestion
```

**Pull request merged**